### PR TITLE
Screens

### PR DESCRIPTION
**Pull request by gigafiga21**
### Changes:
- Added singleton app var to hold screens and manage them
- Added Screen base class to manage widget tree
- Added event-dispatching system

### CODE REVIEW

File: Libraries/Interface/interface.cpp [hunk 1]
```diff
@@ -1,19 +1,172 @@
...
+ * Getter for next widget
+ * @return {Widget*}
+ */
+Widget* Widget::next()
+{
+    return nextWidget;
+}
+
+/**
+ * Setts previous widget
```
Comment (Hemmemm) [inline Libraries/Interface/interface.cpp]: settttttts

File: Libraries/Interface/interface.cpp [hunk 2]
```diff
@@ -1,19 +1,172 @@
...
+ * @param  {unsigned char} name - name of the event
+ * @return {bool}               - true if event listeners are existing and active
+ */
+bool Widget::event(unsigned char name)
+{
+    return events & name;
+}
+
+/**
+ * Setts listener to the event
```
Comment (Hemmemm) [inline Libraries/Interface/interface.cpp]: settttttts

File: Libraries/Interface/interface.cpp [hunk 3]
```diff
@@ -1,19 +1,172 @@
...
+ * @param {Listener}      function - event listener function
+ */
+void Widget::listener(unsigned char name, Listener function)
+{
+    listeners[name].push_back(function);
+    events = events | name;
+}
+
+/**
+ * Dispatches event on widget
```
Comment (Hemmemm) [inline Libraries/Interface/interface.cpp]: Better desription?

File: Libraries/Interface/interface.cpp [hunk 4]
```diff
@@ -1,19 +1,172 @@
...
+ * @param {unsigned char} name        - name of the event
+ */
+void Widget::dispatch(unsigned char name)
+{
+    for (std::list<Listener>::iterator function = listeners[name].begin(); function != listeners[name].end(); function++)
```
Comment (Hemmemm) [inline Libraries/Interface/interface.cpp]: Maybe make some local variables to make code of for statement more readable

File: Libraries/Interface/interface.h [hunk 5]
```diff
@@ -1,16 +1,67 @@
...
+
+/**
+ * Arguments for events methods
+ * @const {unsigned char} EVENT_FOCUS   - event of focusing
+ * @const {unsigned char} EVENT_UNFOCUS - event of unfocusing
+ * @const {unsigned char} EVENT_ACTIVE  - event of activating widget
+ */
+#define EVENT_FOCUS   1
+#define EVENT_UNFOCUS 2
+#define EVENT_ACTIVE  4
```
Comment (Hemmemm) [inline Libraries/Interface/interface.h]: why such numbers? Are they intended to be combined in some way by "|" ?

File: Libraries/Interface/interface.h [hunk 6]
```diff
@@ -1,16 +1,67 @@
...
+typedef std::function<void()> Listener;
 
 /**
  * Basic class of the console widget (to store different widgets in one list)
+ * Contains widgets tree for parsing events by Console class
+ * @prop {Widget*}                                      parentWidget   - parent widget
+ * @prop {Widget*}                                      start          - pointer to the first child of current widget
+ * @prop {Widget*}                                      ending         - pointer to the last child of current widget
+ * @prop {Widget*}                                      nextWidget     - pointer to the previous sibling
+ * @prop {Widget*}                                      previousWidget - pointer to the previous sibling
```
Comment (Hemmemm) [inline Libraries/Interface/interface.h]: both previous? fix

File: Libraries/Interface/menu.cpp [hunk 7]
```diff
@@ -1,12 +1,30 @@
 #include "menu.h"
 
+/**
+ * Draws menu item
+ */
+void MenuItem::draw()
+{
+    move(y, x);
+    printw(name);
```
Comment (Hemmemm) [inline Libraries/Interface/menu.cpp]: Maybe change to addstr?

File: Libraries/Screens/screen.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    return current;
+}
+
+/**
+ * Finds previous focusable widget in tree
+ * @param  {Widget*} start - start point for searching
+ * @return {Widget*}
+ */
+Widget* Screen::previous(Widget* start)
```
Comment (Hemmemm) [inline Libraries/Screens/screen.cpp:79]: Maybe add some explaining comments in the function block

**Review by Hemmemm — CHANGES_REQUESTED** [review]
Ok overall, fix some comments and maybe add some descriptive comments inside function blocks

File: Libraries/Interface/interface.h (hunk 5 above)
Comment (gigafiga21) [reply Libraries/Interface/interface.h]: Yesssss

Comment (Hemmemm) [reply Libraries/Interface/interface.h]: k

File: Libraries/Screens/screen.cpp [hunk 9]
```diff
@@ -64,11 +78,20 @@ Widget* Screen::next(Widget* start)
  */
 Widget* Screen::previous(Widget* start)
 {
+    /**
+     * @var {Widget*} current - start point for searching focusable widget
+     * @var {bool}    bubling - flag for bubbling to the parent of widget (false if previous siblings of widget exist)
+     */
     Widget* current = start;
     char bubbling = true;
```
Comment (Hemmemm) [inline Libraries/Screens/screen.cpp]: make it bool

**Review by Hemmemm — CHANGES_REQUESTED** [review]

**Review by Hemmemm — APPROVED** [review]

**Pull request merged**